### PR TITLE
client/v3: clear auth token when encounter ErrInvalidAuthToken

### LINKED ISSUE #12385: watch stream return "permission denied" if token expired
i ran into the similar problem as #8914

i'm using v3.4.13 server and v3.4.12 client.
i can reproduce the error using following steps:

on one terminal , start etcd server and enable auth and shroten the default token ttl:
```
etcd --log-level warn --logger zap --auth-token-ttl 10 & 
etcdctl user add root:root
etcdctl user grant-role root root
etcdctl auth enable
```

on the other termianal, run a the demo program:
```
cfz@cfz-desktop:~/ws/etcd-watch-demo$ cat main.go
package main

import (
        "context"
        "fmt"
        "sync"
        "time"

        "go.etcd.io/etcd/clientv3"
)

var wg sync.WaitGroup

func main() {
        cfg := clientv3.Config{
                Endpoints: []string{"http://127.0.0.1:2379"},
                Username:  "root",
                Password:  "root",
                LogConfig: nil,
        }

        cli, err := clientv3.New(cfg)
        if err != nil {
                panic(err)
        }

        watch(cli, 1)
        time.Sleep(1 * time.Second)
        watch(cli, 2)
        time.Sleep(11 * time.Second)
        watch(cli, 3)
        time.Sleep(1 * time.Second)
        watch(cli, 4)

        wg.Wait()
}

func watch(cli *clientv3.Client, num int) {
        wg.Add(1)

        go func() {
                defer wg.Done()
                for wr := range cli.Watch(context.Background(), "") {
                        if wr.Err() != nil {
                                fmt.Printf("%#v\n", wr.Err())
                        }
                }

                fmt.Printf("#%v watch existed\n", num)
        }()
}
cfz@cfz-desktop:~/ws/etcd-watch-demo$ go run main.go
&errors.errorString{s:"rpc error: code = PermissionDenied desc = etcdserver: permission denied"}
#3 watch exited
&errors.errorString{s:"rpc error: code = PermissionDenied desc = etcdserver: permission denied"}
#4 watch exited
```

as shown in the log , the 1st and 2nd watch is ok, while the 3rd and 4th report "permission denied"

meanwile, etcd server will log following:
```
{"level":"warn","ts":"2020-10-10T20:34:45.010+0800","caller":"auth/store.go:1301","msg":"invalid auth token","token":"JWourXlCoFPTxMHj.8"}
{"level":"warn","ts":"2020-10-10T20:34:46.010+0800","caller":"auth/store.go:1301","msg":"invalid auth token","token":"JWourXlCoFPTxMHj.8"}
```

i think the causes are:
- the grpc stream's authentication token is provided at the time when the stream is initialized
- the client is trying to reusing the same grpc stream(if avaliable) to setup new watch substream.
- the token attached to the stream would be expired if no other requests to extend its ttl.

### PR DESCRIPTION
**Pull request by bbiao**
Old etcdserver which have not apply pr of #12165 will check auth token
even if the request is an Authenticate request.

If the client has a invalid auth token, it will not able to update it's
token, since the Authenticate has a invalid auth token.
This fix clear the auth token when encounter an ErrInvalidAuthToken to
talk with old version etcd servers.
With pr #12165 and #12264 this will fix the problem in #12385 

Fix #12385

### CODE REVIEW

File: client/v3/retry_interceptor.go [hunk 1]
```diff
@@ -74,6 +74,12 @@ func (c *Client) unaryClientInterceptor(logger *zap.Logger, optFuncs ...retryOpt
 				continue
 			}
 			if callOpts.retryAuth && rpctypes.Error(lastErr) == rpctypes.ErrInvalidAuthToken {
+				// clear auth token befault refreshing it.
+				// call c.Auth.Authenticate with an invalid token will always fail the auth check on the server-side,
+				// if the server has not apply the patch of pr #12165 (https://github.com/etcd-io/etcd/pull/12165)
+				// and a rpctypes.ErrInvalidAuthToken will recursively call c.getToken until system run out of resource.
+				c.authTokenBundle.UpdateAuthToken("")
```
Comment (mitake) [inline client/v3/retry_interceptor.go:81]: Thanks a lot for the fix @bbiao , I think it would work fine but let me make sure it works as expected.

**Review by mitake — APPROVED** [review]
LGTM, thanks a lot for fixing @bbiao ! Defer to @jingyih @gyuho @jpbetz

**Review by cfc4n — APPROVED** [review]
LGTM, Thanks.

File: client/v3/retry_interceptor.go [hunk 2]
```diff
@@ -74,6 +74,12 @@ func (c *Client) unaryClientInterceptor(logger *zap.Logger, optFuncs ...retryOpt
 				continue
 			}
 			if callOpts.retryAuth && rpctypes.Error(lastErr) == rpctypes.ErrInvalidAuthToken {
+				// clear auth token befault refreshing it.
```
Comment (spzala) [inline client/v3/retry_interceptor.go:77]: Nit, typo "befault". Also, don't we need a similar update for https://github.com/etcd-io/etcd/blob/master/client/v3/retry_interceptor.go#L242 for a similar call to  getToken when ErrInvalidAuthToken? Thanks! /cc @mitake @cfc4n

**Comment by spzala** [comment]
@bbiao thanks for addressing my one comment, how about the second comment/question?

**Comment by bbiao** [comment]
> @bbiao thanks for addressing my one comment, how about the second comment/question?

It's ok not to clear auth token when encounter the ErrInvalidAuthToken at https://github.com/etcd-io/etcd/blob/master/client/v3/retry_interceptor.go#L242, since the getToken itself is a unary rpc call, when failed it will retry, and at that moment the unary interceptor will clean the auth token.

But I think it's better to clean the auth token too at https://github.com/etcd-io/etcd/blob/master/client/v3/retry_interceptor.go#L242, this will avoid a never-success rpc call.

**Comment by spzala** [comment]
> > @bbiao thanks for addressing my one comment, how about the second comment/question?
> 
> It's ok not to clear auth token when encounter the ErrInvalidAuthToken at https://github.com/etcd-io/etcd/blob/master/client/v3/retry_interceptor.go#L242, since the getToken itself is a unary rpc call, when failed it will retry, and at that moment the unary interceptor will clean the auth token.
> 
> But I think it's better to clean the auth token too at https://github.com/etcd-io/etcd/blob/master/client/v3/retry_interceptor.go#L242, this will avoid a never-success rpc call.

@bbiao lgtm, thanks for addressing the comments, can you please squash commits now?

**Comment by bbiao** [comment]
@spzala commits have been squashed ;-)

**Comment by mitake** [comment]
Travis is failing because of `linux-amd64-integration-4-cpu` (the case is `TestMemberPromote`) I think this is a non deterministic one. Also semaphore failed because of `TestIssue6361` but on my local env the case could run successfully on the branch. I'll run semaphore again for checking. But I think the PR can be merged.

**Comment by mitake** [comment]
@spzala do you know how to rerun semaphore CI? I found that I cannot do that...

**Comment by spzala** [comment]
@mitake there used to be a rerun option but I also don't see it. Need to investigate it, for now, I think closing/reopening PR is one quick option. As you said, failure seems not related to the changes.

**Comment by spzala** [comment]
@mitake @bbiao the CI is green now :)

**Comment by mitake** [comment]
@spzala yeah that's smart, thanks for handling!

**Pull request merged**